lib/rpmsg/rpmsg_virtio.c

```diff
@@ -638,13 +637,11 @@ int rpmsg_init_vdev(struct rpmsg_virtio_device *rvdev,
 	/* Initialize channels and endpoints list */
 	metal_list_init(&rdev->endpoints);
 
-	dev_features = rpmsg_virtio_get_features(rvdev);
-
 	/*
 	 * Create name service announcement endpoint if device supports name
 	 * service announcement feature.
 	 */
-	if ((dev_features & (1 << VIRTIO_RPMSG_F_NS))) {
+	if (vdev->features & (1 << VIRTIO_RPMSG_F_NS)) {
 		rpmsg_init_ept(&rdev->ns_ept, "NS",
```

Comment:
This modification looks good, just I cannot see where did you set the vdev->features?
From, master, it should set the features before setting the driver_OK, just I didn't see the modification, looks like it should be done in the negotiate_features() API, for slave, it should set after it checks DRIVER_OK. Looks like it needs an virtio API abstraction for it. Or we should have the negotiate_features() API for both roles, but for devices, it will not set the virtio config space, but instead, it should set the features locally, that is the vdev->features.

Reply:
> This modification looks good, just I cannot see where did you set the vdev->features?
In rproc_virtio_create_vdev, I add one line to set it:
vdev->features = rproc_virtio_get_features(vdev);

> From, master, it should set the features before setting the driver_OK, just I didn't see the modification, looks like it should be done in the negotiate_features() API,
From virtio spec, master should:
1.Read dfeature field from resource table
2.Mask the feature which doesn't support
3.Write the result to gfeature field
The current implementation doesn't confirm the spec, do you want to fully implement the negotiation process.

>for slave, it should set after it checks DRIVER_OK. Looks like it needs an virtio API abstraction for it. Or we should have the negotiate_features() API for both roles, but for devices, it will not set the virtio config space, but instead, it should set the features locally, that is the vdev->features.
No, the slave side shouldn't set any field in config space, instead it should get the negotiated result from gfeature directly.

Reply:
@wjliang I add new patch to negotiate the feature, please review.

Reply:
> No, the slave side shouldn't set any field in config space, instead it should get the negotiated result from gfeature directly.

What i mean by "set" is to set the local features field which is the negotiation result. In another commit of the same pull request, i have comments in the negotiate_features.

Just for get_features for master, it should not just get the dfeatures, it should be the negotiated result. It looks like based on your modification, gfeatures sounds like the result, but i am not sure if this is always the case in future, maybe better to do dfeatures & gfeatures in the get_features() function.

Reply:
Ok, I will change to dfeatures & gfeatures.